File: contribai/generator/scorer.py
"""Contribution quality scorer.

Evaluates generated contributions before submission
to prevent low-quality PRs from being created.
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass

from contribai.core.models import Contribution, ContributionType

logger = logging.getLogger(__name__)


@dataclass
class QualityReport:
    """Quality assessment of a contribution."""

    score: float  # 0.0 - 1.0
    passed: bool
    checks: dict[str, CheckResult]

    @property
    def summary(self) -> str:
        passed = sum(1 for c in self.checks.values() if c.passed)
        total = len(self.checks)
        return f"{passed}/{total} checks passed (score: {self.score:.0%})"


@dataclass
class CheckResult:
    """Result of a single quality check."""

    name: str
    passed: bool
    score: float  # 0.0 - 1.0
    reason: str


class QualityScorer:
    """Evaluates contribution quality before PR submission.

    Runs a series of heuristic checks to catch low-quality
    contributions that would likely be rejected by maintainers.
    """

    def __init__(self, min_score: float = 0.6):
        self._min_score = min_score

    def evaluate(self, contribution: Contribution) -> QualityReport:
        """Run all quality checks on a contribution.

        Args:
            contribution: The generated contribution to evaluate.

        Returns:
            QualityReport with individual check results and overall score.
        """
        checks = {}

        checks["has_changes"] = self._check_has_changes(contribution)
        checks["change_size"] = self._check_change_size(contribution)
        checks["commit_message"] = self._check_commit_message(contribution)
        checks["description"] = self._check_description(contribution)
        checks["no_debug_code"] = self._check_no_debug_code(contribution)
        checks["no_placeholders"] = self._check_no_placeholders(contribution)
        checks["file_coherence"] = self._check_file_coherence(contribution)

        total_score = sum(c.score for c in checks.values()) / len(checks)
        passed = total_score >= self._min_score

        report = QualityReport(score=total_score, passed=passed, checks=checks)
        logger.info("Quality check: %s", report.summary)
        return report

    def _check_has_changes(self, c: Contribution) -> CheckResult:
        """At least one meaningful file change."""
        has = len(c.changes) > 0 and any(
            len(ch.new_content.strip()) > 0 for ch in c.changes
        )
        return CheckResult(
            name="has_changes",
            passed=has,
            score=1.0 if has else 0.0,
            reason="Has file changes" if has else "No file changes",
        )

    def _check_change_size(self, c: Contribution) -> CheckResult:
        """Changes should be focused (not too big, not trivial)."""
        total_lines = sum(len(ch.new_content.splitlines()) for ch in c.changes)

        if total_lines == 0:
            return CheckResult("change_size", False, 0.0, "Empty changes")
        elif total_lines < 3:
            return CheckResult(
                "change_size", False, 0.3,
                f"Very small change ({total_lines} lines)",
            )
        elif total_lines > 500:
            return CheckResult(
                "change_size", False, 0.4,
                f"Very large change ({total_lines} lines)",
            )
        elif total_lines > 200:
            return CheckResult("change_size", True, 0.7, f"Large change ({total_lines} lines)")
        else:
            return CheckResult("change_size", True, 1.0, f"Good change size ({total_lines} lines)")

    def _check_commit_message(self, c: Contribution) -> CheckResult:
        """Commit message follows conventional format."""
        msg = c.commit_message
        if not msg:
            return CheckResult("commit_message", False, 0.0, "Empty commit message")

        # Check conventional commit format: type: description
        conventional = re.match(r'^(feat|fix|docs|refactor|perf|test|chore)\(?.*\)?: .+', msg)
        if conventional:
            return CheckResult("commit_message", True, 1.0, "Follows conventional commits")

        if len(msg) > 10:
            return CheckResult("commit_message", True, 0.7, "Descriptive but not conventional")

        return CheckResult("commit_message", False, 0.3, "Poor commit message")

    def _check_description(self, c: Contribution) -> CheckResult:
        """PR description is meaningful."""
        desc = c.description
        if not desc:
            return CheckResult("description", False, 0.0, "Empty description")
        if len(desc) < 20:
            return CheckResult("description", False, 0.3, "Description too short")
        return CheckResult("description", True, 1.0, "Good description")

    def _check_no_debug_code(self, c: Contribution) -> CheckResult:
        """No debug statements in generated code."""
        debug_patterns = [
            r'\bprint\s*\(',
            r'\bconsole\.log\s*\(',
            r'\bdebugger\b',
            r'\bpdb\.set_trace\(',
            r'\bbreakpoint\(\)',
            r'#\s*TODO\b',
            r'#\s*FIXME\b',
            r'#\s*HACK\b',
        ]

        issues = []
        for change in c.changes:
            for pattern in debug_patterns:
                if re.search(pattern, change.new_content):
                    issues.append(f"{change.path}: {pattern}")

        if not issues:
            return CheckResult("no_debug_code", True, 1.0, "No debug code found")

        # Allow some patterns (TODO can be intentional)
        severity = 0.8 if len(issues) <= 2 else 0.4
        return CheckResult(
            "no_debug_code",
            severity >= 0.6,
            severity,
            f"Found {len(issues)} debug patterns",
        )

    def _check_no_placeholders(self, c: Contribution) -> CheckResult:
        """No placeholder text in generated code."""
        placeholder_patterns = [
            r'YOUR_.*_HERE',
            r'REPLACE_THIS',
            r'PLACEHOLDER',
            r'XXX',
            r'lorem ipsum',
            r'example\.com',
            r'foo\s*bar',
        ]

        for change in c.changes:
            content_lower = change.new_content.lower()
            for pattern in placeholder_patterns:
                if re.search(pattern, content_lower, re.IGNORECASE):
                    return CheckResult(
                        "no_placeholders", False, 0.2,
                        f"Found placeholder: {pattern} in {change.path}",
                    )

        return CheckResult("no_placeholders", True, 1.0, "No placeholders found")

    def _check_file_coherence(self, c: Contribution) -> CheckResult:
        """Changes are related to the finding."""
        if not c.changes:
            return CheckResult("file_coherence", False, 0.0, "No changes")

        # Check that the finding's file is actually changed
        finding_file = c.finding.file_path
        changed_files = {ch.path for ch in c.changes}

        if finding_file in changed_files:
            return CheckResult("file_coherence", True, 1.0, "Finding file is changed")

        # Some contributions legitimately change different files
        if c.contribution_type in (ContributionType.DOCS_IMPROVE, ContributionType.FEATURE_ADD):
            return CheckResult("file_coherence", True, 0.8, "Different file but type allows it")

        return CheckResult(
            "file_coherence", False, 0.4,
            f"Finding in {finding_file} but changes in {changed_files}",
        )

File: crates/contribai-rs/src/generator/scorer.rs
//! Contribution quality scorer.
//!
//! Port from Python `generator/scorer.py`.
//! Evaluates generated contributions before submission
//! to prevent low-quality PRs.

use regex::Regex;
use std::sync::LazyLock;
use tracing::info;

use crate::core::models::Contribution;

static RE_CONVENTIONAL_COMMIT: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^(feat|fix|docs|refactor|perf|test|chore)\(?.*\)?: .+").unwrap());

/// Quality assessment of a contribution.
#[derive(Debug, Clone)]
pub struct QualityReport {
    pub score: f64, // 0.0 - 1.0
    pub passed: bool,
    pub checks: Vec<CheckResult>,
}

impl QualityReport {
    pub fn summary(&self) -> String {
        let passed = self.checks.iter().filter(|c| c.passed).count();
        format!(
            "{}/{} checks passed (score: {:.0}%)",
            passed,
            self.checks.len(),
            self.score * 100.0
        )
    }
}

/// Result of a single quality check.
#[derive(Debug, Clone)]
pub struct CheckResult {
    pub name: String,
    pub passed: bool,
    pub score: f64, // 0.0 - 1.0
    pub reason: String,
}

/// Evaluates contribution quality before PR submission.
pub struct QualityScorer {
    min_score: f64,
}

impl QualityScorer {
    pub fn new(min_score: f64) -> Self {
        Self { min_score }
    }

    /// Run all quality checks on a contribution.
    pub fn evaluate(
        &self,
        contribution: &Contribution,
        repo_prefs: Option<&crate::orchestrator::memory::RepoPreferences>,
    ) -> QualityReport {
        let checks = vec![
            self.check_has_changes(contribution),
            self.check_change_size(contribution),
            self.check_commit_message(contribution),
            self.check_description(contribution),
            self.check_no_debug_code(contribution),
            self.check_no_placeholders(contribution),
            self.check_file_coherence(contribution),
            self.check_outcome_history(contribution, repo_prefs),
        ];

        let total_score: f64 = checks.iter().map(|c| c.score).sum::<f64>() / checks.len() as f64;
        let passed = total_score >= self.min_score;

        let report = QualityReport {
            score: total_score,
            passed,
            checks,
        };

        info!(summary = %report.summary(), "Quality check");
        report
    }

    fn check_has_changes(&self, c: &Contribution) -> CheckResult {
        let has =
            !c.changes.is_empty() && c.changes.iter().any(|ch| !ch.new_content.trim().is_empty());
        CheckResult {
            name: "has_changes".into(),
            passed: has,
            score: if has { 1.0 } else { 0.0 },
            reason: if has {
                "Has file changes".into()
            } else {
                "No file changes".into()
            },
        }
    }

    fn check_change_size(&self, c: &Contribution) -> CheckResult {
        let total_lines: usize = c
            .changes
            .iter()
            .map(|ch| ch.new_content.lines().count())
            .sum();

        if total_lines == 0 {
            CheckResult {
                name: "change_size".into(),
                passed: false,
                score: 0.0,
                reason: "Empty changes".into(),
            }
        } else if total_lines < 3 {
            CheckResult {
                name: "change_size".into(),
                passed: false,
                score: 0.3,
                reason: format!("Very small change ({} lines)", total_lines),
            }
        } else if total_lines > 500 {
            CheckResult {
                name: "change_size".into(),
                passed: false,
                score: 0.4,
                reason: format!("Very large change ({} lines)", total_lines),
            }
        } else if total_lines > 200 {
            CheckResult {
                name: "change_size".into(),
                passed: true,
                score: 0.7,
                reason: format!("Large change ({} lines)", total_lines),
            }
        } else {
            CheckResult {
                name: "change_size".into(),
                passed: true,
                score: 1.0,
                reason: format!("Good change size ({} lines)", total_lines),
            }
        }
    }

    fn check_commit_message(&self, c: &Contribution) -> CheckResult {
        let msg = &c.commit_message;
        if msg.is_empty() {
            return CheckResult {
                name: "commit_message".into(),
                passed: false,
                score: 0.0,
                reason: "Empty commit message".into(),
            };
        }

        if RE_CONVENTIONAL_COMMIT.is_match(msg) {
            CheckResult {
                name: "commit_message".into(),
                passed: true,
                score: 1.0,
                reason: "Follows conventional commits".into(),
            }
        } else if msg.len() > 10 {
            CheckResult {
                name: "commit_message".into(),
                passed: true,
                score: 0.7,
                reason: "Descriptive but not conventional".into(),
            }
        } else {
            CheckResult {
                name: "commit_message".into(),
                passed: false,
                score: 0.3,
                reason: "Poor commit message".into(),
            }
        }
    }

    fn check_description(&self, c: &Contribution) -> CheckResult {
        if c.description.is_empty() {
            CheckResult {
                name: "description".into(),
                passed: false,
                score: 0.0,
                reason: "Empty description".into(),
            }
        } else if c.description.len() < 20 {
            CheckResult {
                name: "description".into(),
                passed: false,
                score: 0.3,
                reason: "Description too short".into(),
            }
        } else {
            CheckResult {
                name: "description".into(),
                passed: true,
                score: 1.0,
                reason: "Good description".into(),
            }
        }
    }

    fn check_no_debug_code(&self, c: &Contribution) -> CheckResult {
        let debug_patterns = [
            r"\bprint\s*\(",
            r"\bconsole\.log\s*\(",
            r"\bdebugger\b",
            r"\bpdb\.set_trace\(",
            r"\bbreakpoint\(\)",
            r"#\s*TODO\b",
            r"#\s*FIXME\b",
            r"#\s*HACK\b",
        ];

        for change in &c.changes {
            for pattern in &debug_patterns {
                if let Ok(re) = Regex::new(pattern) {
                    if re.is_match(&change.new_content) {
                        return CheckResult {
                            name: "no_debug_code".into(),
                            passed: false,
                            score: 0.3,
                            reason: format!("Debug code found: {}", pattern),
                        };
                    }
                }
            }
        }

        CheckResult {
            name: "no_debug_code".into(),
            passed: true,
            score: 1.0,
            reason: "No debug code found".into(),
        }
    }

    fn check_no_placeholders(&self, c: &Contribution) -> CheckResult {
        let placeholder_patterns = [
            "TODO: implement",
            "PLACEHOLDER",
            "YOUR_CODE_HERE",
            "pass  # TODO",
            "raise NotImplementedError",
            "...",
        ];

        for change in &c.changes {
            for pattern in &placeholder_patterns {
                if change.new_content.contains(pattern) {
                    return CheckResult {
                        name: "no_placeholders".into(),
                        passed: false,
                        score: 0.2,
                        reason: format!("Placeholder found: {}", pattern),
                    };
                }
            }
        }

        CheckResult {
            name: "no_placeholders".into(),
            passed: true,
            score: 1.0,
            reason: "No placeholders found".into(),
        }
    }

    fn check_outcome_history(
        &self,
        c: &Contribution,
        prefs: Option<&crate::orchestrator::memory::RepoPreferences>,
    ) -> CheckResult {
        let Some(prefs) = prefs else {
            return CheckResult {
                name: "outcome_history".into(),
                passed: true,
                score: 0.7,
                reason: "No outcome history available".into(),
            };
        };

        let type_str = format!("{:?}", c.contribution_type).to_lowercase();

        if prefs
            .rejected_types
            .iter()
            .any(|t| t.to_lowercase() == type_str)
        {
            return CheckResult {
                name: "outcome_history".into(),
                passed: false,
                score: 0.2,
                reason: format!("Type '{}' was previously rejected by this repo", type_str),
            };
        }

        if prefs
            .preferred_types
            .iter()
            .any(|t| t.to_lowercase() == type_str)
        {
            return CheckResult {
                name: "outcome_history".into(),
                passed: true,
                score: 1.0,
                reason: format!("Type '{}' was previously merged by this repo", type_str),
            };
        }

        let (score, reason) = if prefs.merge_rate < 0.2 {
            (0.4, "Repo has low merge rate (<20%)")
        } else if prefs.merge_rate >= 0.5 {
            (0.9, "Repo has good merge rate (>=50%)")
        } else {
            (0.6, "Repo has moderate merge rate")
        };

        CheckResult {
            name: "outcome_history".into(),
            passed: score >= 0.5,
            score,
            reason: reason.into(),
        }
    }

    fn check_file_coherence(&self, c: &Contribution) -> CheckResult {
        // Check that all changed files relate to the finding
        let finding_dir = c
            .finding
            .file_path
            .rsplit_once('/')
            .map(|(dir, _)| dir)
            .unwrap_or("");

        let unrelated = c
            .changes
            .iter()
            .filter(|ch| {
                let change_dir = ch.path.rsplit_once('/').map(|(d, _)| d).unwrap_or("");
                !change_dir.starts_with(finding_dir) && !finding_dir.starts_with(change_dir)
            })
            .count();

        if unrelated > 0 {
            CheckResult {
                name: "file_coherence".into(),
                passed: false,
                score: 0.5,
                reason: format!("{} unrelated file(s) changed", unrelated),
            }
        } else {
            CheckResult {
                name: "file_coherence".into(),
                passed: true,
                score: 1.0,
                reason: "All changes are related".into(),
            }
        }
    }
}

impl Default for QualityScorer {
    fn default() -> Self {
        Self::new(0.6)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::models::{ContributionType, FileChange, Finding, Severity};
    use chrono::Utc;

    fn test_contribution() -> Contribution {
        Contribution {
            finding: Finding {
                id: "1".into(),
                finding_type: ContributionType::SecurityFix,
                severity: Severity::High,
                title: "SQL injection".into(),
                description: "User input is not sanitized before use in SQL queries".into(),
                file_path: "src/db/queries.py".into(),
                line_start: Some(10),
                line_end: Some(15),
                suggestion: Some("Use parameterized queries".into()),
                confidence: 0.9,
                priority_signals: vec![],
            },
            contribution_type: ContributionType::SecurityFix,
            title: "fix: sql injection in user query".into(),
            description: "User input is not sanitized before use in SQL queries".into(),
            changes: vec![FileChange {
                path: "src/db/queries.py".into(),
                original_content: None,
                new_content: "def query(user_id):\n    cursor.execute('SELECT * FROM users WHERE id = %s', (user_id,))\n".into(),
                is_new_file: false,
                is_deleted: false,
            }],
            commit_message: "fix(db): sanitize sql query parameters".into(),
            tests_added: vec![],
            branch_name: "fix/security/sql-injection".into(),
            generated_at: Utc::now(),
        }
    }

    #[test]
    fn test_quality_scorer_good_contribution() {
        let scorer = QualityScorer::default();
        let c = test_contribution();
        let report = scorer.evaluate(&c, None);

        assert!(
            report.passed,
            "Good contribution should pass: {}",
            report.summary()
        );
        assert!(report.score >= 0.6);
    }

    #[test]
    fn test_quality_scorer_empty_changes() {
        let scorer = QualityScorer::default();
        let mut c = test_contribution();
        c.changes = vec![];

        let report = scorer.evaluate(&c, None);
        let check = report
            .checks
            .iter()
            .find(|c| c.name == "has_changes")
            .unwrap();
        assert!(!check.passed, "has_changes should fail with no changes");
        assert_eq!(check.score, 0.0);
    }

    #[test]
    fn test_quality_scorer_debug_code() {
        let scorer = QualityScorer::default();
        let mut c = test_contribution();
        c.changes = vec![FileChange {
            path: "src/db/queries.py".into(),
            original_content: None,
            new_content: "print('debugging here')\ndef query():\n    pass\n".into(),
            is_new_file: false,
            is_deleted: false,
        }];

        let report = scorer.evaluate(&c, None);
        let debug_check = report.checks.iter().find(|c| c.name == "no_debug_code");
        assert!(!debug_check.unwrap().passed, "Should catch debug code");
    }

    #[test]
    fn test_quality_scorer_conventional_commit() {
        let scorer = QualityScorer::default();
        let c = test_contribution();
        let report = scorer.evaluate(&c, None);

        let commit_check = report.checks.iter().find(|c| c.name == "commit_message");
        assert!(commit_check.unwrap().passed);
        assert_eq!(commit_check.unwrap().score, 1.0);
    }

    #[test]
    fn test_outcome_history_no_prefs_neutral() {
        let scorer = QualityScorer::default();
        let c = test_contribution();
        let report = scorer.evaluate(&c, None);
        let check = report
            .checks
            .iter()
            .find(|c| c.name == "outcome_history")
            .unwrap();
        assert!(check.passed);
        assert!((check.score - 0.7).abs() < 0.01);
    }

    #[test]
    fn test_outcome_history_rejected_type_penalized() {
        use crate::orchestrator::memory::RepoPreferences;
        let scorer = QualityScorer::default();
        let c = test_contribution();
        let prefs = RepoPreferences {
            preferred_types: vec![],
            rejected_types: vec!["securityfix".into()],
            merge_rate: 0.5,
            avg_review_hours: 24.0,
            notes: String::new(),
        };
        let report = scorer.evaluate(&c, Some(&prefs));
        let check = report
            .checks
            .iter()
            .find(|c| c.name == "outcome_history")
            .unwrap();
        assert!(!check.passed);
        assert!((check.score - 0.2).abs() < 0.01);
    }

    #[test]
    fn test_outcome_history_preferred_type_boosted() {
        use crate::orchestrator::memory::RepoPreferences;
        let scorer = QualityScorer::default();
        let c = test_contribution();
        let prefs = RepoPreferences {
            preferred_types: vec!["securityfix".into()],
            rejected_types: vec![],
            merge_rate: 0.5,
            avg_review_hours: 24.0,
            notes: String::new(),
        };
        let report = scorer.evaluate(&c, Some(&prefs));
        let check = report
            .checks
            .iter()
            .find(|c| c.name == "outcome_history")
            .unwrap();
        assert!(check.passed);
        assert!((check.score - 1.0).abs() < 0.01);
    }
}
